src/Foundation/NSDate.cs

```diff
@@ -65,42 +74,20 @@ public partial class NSDate {
 
 			var dtUnv = dt.ToUniversalTime ();
 
-			var calendar = new NSCalendar (NSCalendarType.Gregorian);
-			calendar.TimeZone = NSTimeZone.FromName ("UTC");
-			var components = new NSDateComponents {
-				Day = dtUnv.Day, Month = dtUnv.Month, Year = dtUnv.Year, Hour = dtUnv.Hour,
-					Minute = dtUnv.Minute, Second = dtUnv.Second, Nanosecond = dtUnv.Millisecond * 1000000
-			};
+			components.Day = dtUnv.Day;
```

Comment:
This can race is NSDate is used from multiple threads?

Reply:
One trick that can be used, if allocations need to be avoided on this path is using a `ThreadLocal<NSDate>`

Reply:
Correct. The components cannot be moved to a static variable. Since several threads can be executing from lines 77 to 83 interrupting each other, making the components to be useless. Now, we can move the declaration of the NSComponents in the static variable to:

```chsarp
private static readonly ThreadLocal<NSDateComponents> components = new ThreadLocal<NSDateComponents>(() => new NSDateComponents());
```

At that point we have a NSDateComponents value per thread, and should not be stepping on each other. Now, and just for a reference, something similar can be achieved with the ThreadStatic attribute  https://docs.microsoft.com/en-us/dotnet/api/system.threadstaticattribute?view=netframework-4.8 

```csharp
[ThreadStatic] static NSDateComponents components = new NSDateComponents ();
```

But the devil is in the detail (or so I have been told). ThreadStatic will always create the object, while ThreadLocal will only be calling the delegate once the value is needed, that is, is lazy, and in this context we want this. If we were using ThreadStatic, we will ALWAYS create the NSDateComponents, even when is not needed. When using ThreadLocal, the evaluation of the callback will only happen in the first call of the Value, so if the user never uses the cast, the object is not created.

Reply:
So, ThreadStatic is going to be icky. With ThreadStatic you need to do an assignment to the variable in each thread.

So effectively, all usages of `components` will have to be like: `components = components ?? new NSDateComponents()` as the initializer is only called for the first running thread.

Honestly, ThreadLocal is the simplest and safest way forward.

Reply:
Just to clarify, I added the ThreadStatic explanation to give background on why I'm on board with ThreadLocal. Much better approach.